File: implementation-go/wallet-service/service.go
package wallet_service

import (
	"context"
	"fmt"
	"go-microtrans-grpc/ledger-service/pb/ledger"

	"go-microtrans-grpc/user-service/pb/user"
	"go-microtrans-grpc/wallet-service/pb/wallet"
)

type WalletService struct {
	userClient   user.UserServiceClient
	ledgerClient ledger.LedgerServiceClient
}

func (s *WalletService) checkEligibility(ctx context.Context, userId int64, amount float64) (*user.UserEligibilityResponse, error) {
	return s.userClient.ValidateUser(ctx, &user.UserEligibilityRequest{
		UserId:        userId,
		AmountToSpend: amount,
	})
}

func (s *WalletService) updateBalance(ctx context.Context, userId int64, amount float64, isCredit bool) (*user.UpdateBalanceResponse, error) {
	return s.userClient.UpdateUserBalance(ctx, &user.UpdateBalanceRequest{
		UserId:   userId,
		Amount:   amount,
		IsCredit: isCredit,
	})
}

func (s *WalletService) recordToLedger(ctx context.Context, transactionType int32, userId int64, amount float64, status bool) (*ledger.RecordTransactionResponse, error) {
	return s.ledgerClient.RecordTransaction(ctx, &ledger.RecordTransactionRequest{
		TransactionType: transactionType,
		UserId:          userId,
		Amount:          amount,
		Status:          status,
	})
}

func (s *WalletService) ExecuteTransaction(ctx context.Context, req *wallet.TransactionRequest) (*wallet.TransactionResponse, error) {
	userId := req.GetUserId()
	amount := req.GetAmount()
	transactionType := req.GetTransactionType()

	userCheck, err := s.checkEligibility(ctx, userId, amount)
	if err != nil {
		return nil, err
	}
	if !userCheck.GetIsValid() {
		return &wallet.TransactionResponse{
			Success: false,
			Message: fmt.Sprintf("Transaction failed due to user not found"),
		}, nil
	}

	isDebit := TransactionType(transactionType) != TopUp
	if isDebit && !userCheck.GetBalanceSufficient() {
		return &wallet.TransactionResponse{
			Success: false,
			Message: fmt.Sprintf("Transaction failed due to insufficient balance "),
		}, nil
	}

	updateRes, err := s.updateBalance(ctx, userId, amount, !isDebit)
	if err != nil {
		return nil, err
	}

	if !updateRes.GetStatus() {
		return &wallet.TransactionResponse{
			Success: false,
			Message: fmt.Sprintf("Transaction failed: %s", updateRes.GetMessage()),
		}, nil
	}

	ledgerRes, err := s.recordToLedger(ctx, transactionType, userId, amount, updateRes.GetStatus())
	if err != nil {
		return nil, err
	}

	return &wallet.TransactionResponse{
		Success:       true,
		Message:       "Transaction succeeded",
		TransactionId: ledgerRes.GetReferenceId(),
		FinalBalance:  updateRes.GetCurrentBalance(),
	}, nil
}

func NewWalletService(userClient user.UserServiceClient, ledgerClient ledger.LedgerServiceClient) *WalletService {
	return &WalletService{
		userClient:   userClient,
		ledgerClient: ledgerClient,
	}
}

File: implementation-nodejs/wallet-service/src/service.js
const { TransactionType } = require("./constant");

class WalletService 
{
    constructor(userServiceClient, ledgerServiceClient) 
    {
        this.userClient = userServiceClient;
        this.ledgerClient = ledgerServiceClient;
    }

    checkEligibility(userId, amount) 
    {
        return new Promise((resolve, reject) => {
            this.userClient.ValidateUser({ user_id: userId, amount_to_spend: amount }, (err, response) => {
                if (err) return reject(err);
                resolve(response);
            });
        });
    }

    updateBalance(userId, amount, isCredit) 
    {
        return new Promise((resolve, reject) => {
            this.userClient.UpdateUserBalance({ user_id: userId, amount: amount, is_credit: isCredit }, (err, response) => {
                if (err) return reject(err);
                resolve(response);
            });
        });
    }

    recordToLedger(transactionType, userId, amount, status) 
    {
        return new Promise((resolve, reject) => {
            this.ledgerClient.RecordTransaction({
                transaction_type: transactionType,
                user_id: userId,
                amount: amount,
                status: status
            }, (err, response) => {
                if (err) return reject(err);
                resolve(response);
            });
        });
    }

    async executeTransaction(userId, amount, transactionType) 
    {
        try 
        {
            // 1. Check Eligibility
            const userCheck = await this.checkEligibility(userId, amount);

            if (!userCheck.is_valid) 
            {
                return {
                    success: false,
                    message: "Transaction failed due to user not found",
                    transaction_id: "",
                    final_balance: 0
                };
            }

            const isDebit = transactionType !== TransactionType.TopUp;

            if (isDebit && !userCheck.balance_sufficient) 
            {
                return {
                    success: false,
                    message: "Transaction failed due to insufficient balance ",
                    transaction_id: "",
                    final_balance: 0
                };
            }

            // 2. Update Balance
            const updateRes = await this.updateBalance(userId, amount, !isDebit);

            if (!updateRes.status) 
            {
                return {
                    success: false,
                    message: `Transaction failed: ${updateRes.message}`,
                    transaction_id: "",
                    final_balance: 0
                };
            }

            // 3. Record to Ledger
            const ledgerRes = await this.recordToLedger(transactionType, userId, amount, updateRes.status);

            return {
                success: true,
                message: "Transaction succeeded",
                transaction_id: ledgerRes.reference_id,
                final_balance: updateRes.current_balance
            };

        } 
        catch (error) 
        {
            console.error("ExecuteTransaction error:", error);
            throw new Error(error.message);
        }
    }
}

module.exports = WalletService;